tool/src/java/mod.rs

```diff
@@ -0,0 +1,1667 @@
+    domain: Cow<'a, str>,
+    edges: Vec<Cow<'a, str>>,
+    static_methods: Vec<Cow<'a, str>>,
+    class_methods: Vec<Cow<'a, str>>,
+}
+
+#[derive(Debug, Clone)]
+struct LifetimeTpl<'a> {
+    name: Cow<'a, str>,
+    edges: Vec<Cow<'a, str>>,
+}
+
+#[derive(Template, Clone, Debug)]
+#[template(path = "java/StructConversion.java.jinja", escape = "none")]
+pub(crate) struct StructConversionTpl<'a> {
+    lifetimes: Vec<LifetimeTpl<'a>>,
+    ty: Cow<'a, str>,
+    name: Cow<'a, str>,
+    native_val: Cow<'a, str>,
+}
+
+#[derive(Template, Clone, Debug)]
+#[template(path = "java/OpaqueConversion.java.jinja", escape = "none")]
+pub(crate) struct OpaqueConversionTpl<'a> {
+    name: Cow<'a, str>,
+    lifetimes: Vec<LifetimeTpl<'a>>,
+    self_edges: Option<Cow<'a, str>>,
+    ty: Cow<'a, str>,
+    native_val: Cow<'a, str>,
+}
+
+#[derive(Template, Clone, Debug)]
+#[template(path = "java/OpaqueConstructor.java.jinja", escape = "none")]
+pub(crate) struct OpaqueConstructorTpl<'a> {
+    return_ty: Cow<'a, str>,
+}
+
+#[derive(Debug, Clone, Deserialize)]
+struct Config<'cx> {
+    domain: Cow<'cx, str>,
+    lib_name: Cow<'cx, str>,
+}
+
+struct TyGenContext<'a, 'cx> {
+    tcx: &'cx TypeContext,
+    tcx_config: Config<'cx>,
+    formatter: &'a JavaFormatter<'cx>,
+    errors: &'a ErrorStore<'cx, String>,
+}
+
+#[derive(Debug, Clone)]
+struct Conversion<'cx> {
+    conversion: Cow<'cx, str>,
+    converted_value: Cow<'cx, str>,
+}
+
+mod arena {
+    use std::fmt::Display;
+    pub type Render<'cx> = Box<dyn Fn(Arena<'cx>) -> Conversion<'cx> + 'cx>;
+
+    const UNIT: &() = &();
+    use super::Conversion;
+
+    #[must_use]
+    pub(crate) struct AllocationConversion<'cx, F: Fn(Arena<'cx>) -> Conversion<'cx>> {
+        clos: F,
+        _marker: &'cx (),
+    }
+    impl<'cx, F: Fn(Arena<'cx>) -> Conversion<'cx> + 'cx> AllocationConversion<'cx, F> {
+        pub fn dynamic(clos: F) -> AllocationConversion<'cx, Render<'cx>> {
+            AllocationConversion {
+                clos: Box::new(clos) as Render<'cx>,
+                _marker: UNIT,
+            }
+        }
+
+        pub fn render(self, arena: Arena<'cx>) -> Conversion<'cx> {
+            (self.clos)(arena)
+        }
+    }
+
+    pub enum Arena<'cx> {
+        Closed,
+        Auto,
+        #[allow(unused)]
+        Phantom(&'cx ()),
+    }
+
+    impl<'cx> Display for Arena<'cx> {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            match self {
+                Arena::Closed => write!(f, "arena"),
+                Arena::Auto => write!(f, "Arena.ofAuto()"),
+                Arena::Phantom(_) => write!(f, "phantom"),
+            }
+        }
+    }
+}
+
+enum AllocateConversion<'cx, F: Fn(arena::Arena<'cx>) -> Conversion<'cx>> {
+    NotAllocating(Conversion<'cx>),
+    Allocating(arena::AllocationConversion<'cx, F>),
+}
+
+impl<'a, 'cx> TyGenContext<'a, 'cx> {
+    #[allow(unused)]
+    fn java_to_native<T: TyPosition>(
```

Comment:
As per your comment last review @Manishearth I tried to harmonise the param conversion to just a language conversion, so it can be reused in e.g. the struct conversion. It feels a lot cleaner but I admit the control inversion above is a little weird.